framework/include/materials/MaterialData.h

```diff
@@ -79,6 +79,22 @@ class MaterialData
   /// copy material properties from one element to another
   void copy(const Elem & elem_to, const Elem & elem_from, unsigned int side);
 
+  /**
+   * Copy material properties to elem_to.
+   *
+   * @param elem_to Element to copy data to
+   * @param props material property data to copy
+   * @param props_old old material property data to copy
+   * @param props_older older material property data to copy, can be empty if the target
+   * element has no older properties
+   * @param side Side number (elemental material properties have this equal to zero)
+   */
+  void copy(const Elem & elem_to,
+            const HashMap<unsigned int, MaterialProperties> & props,
+            const HashMap<unsigned int, MaterialProperties> & props_old,
+            const HashMap<unsigned int, MaterialProperties> & props_older,
+            unsigned int side);
```

Comment:
Just to satisfy my curiosity, why do you guys need to manually pass in the hashmaps here?  Why can't you just pass in a material data object like the other copy methods do?

Reply:
Or actually just pass in nothing (since this is a material data object) - and use the maps from the material data's mat prop storage object?

Reply:
I think at the time we want to copy the material data, the elem_from no longer exists. 

At least that's why I didn't do what you suggested. Maybe you can help us think of a way around, so let me explain to you what we are trying to do:

1. First, there is this `element A`.
2. At some point, we want introduce a cut. So `element A` is cut with XFEM into `partial element B` and `partial element C`. `element A` is deleted.
3. Then, we want to update the location or orientation of the cut, so we first heal the cut. To do that, we delete `partial element B`, and heal `partial element C` to get `element D`.
4. Next, `element D` is re-cut with updated location and orientation of the cut, into `partial element E` and `partial element F`.

Then, obviously partial elements E and F lose the stateful information from partial elements B and C. This PR attempts to fix that. Before healing the cut in step 3, stateful properties from partial elements B and C are stored into a data structure we maintain in XFEM. After re-cutting the element in step 4, previously stored stateful data are copied into partial elements E and F using this new API in MaterialData we added.

With all that being said, it is still possible to avoid extracting and storing material properties from partial elements B and C if we can make them "live longer". In other words, if partial elements B and C are not deleted until we have created partial elements E and F, we can just use the old API to copy around the material properties without introducing this new API. I think @bwspenc can comment on whether this is possible without an overhaul of the current implementation of XFEM.

Reply:
Do steps 1 through 4 all happen on the same processor - or could there be a repartitioning or other such elem-moves between any of the steps?  Stateful material properties are stored in a separate data structure - not directly on/with the elements themselves.  So when you delete an element - its stateful property values should still be sitting inside the stateful prop storage data structures.  They are indexed by element pointers.  The problem you cited is that the element object may have been deleted - but it's previous pointer value should still work for indexing those data structures.  The real problem is that our existing APIs for copying/swapping, etc. take element references - when in actuality all they ever do is address the elem to get its underlying pointer/address.  I suspect that if we were to modify the existing APIs, you might not even need to copy around those material data objects - as long as the stateful properties had been synced to the main/sim-global storage loc by the time you want to do this copying.

Reply:
Thanks, @rwcarlsen, we'll do what you are suggesting!  That will considerably simplify things on our end.

Reply:
I suppose, however, that there is the risk that element deleting/creating could create a new element pointer with the same memory address as one of the previous elements - an address that was used to index a stateful property.  However, this would only be a problem if in the interrim, stateful properties associated with the new element reusing the old address were stored.  Seems really unlikely to affect you guys.  But this sort of thing is already a problem with our pointer-indexing as it is.  We really should use elem ids instead.  Someday we'll have time to redo all the material prop stuff.